src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java

```diff
@@ -120,143 +203,208 @@ private static IVecInt toVecInt(IntStream stream) {
+						ModLink link = iterator.next();
+						if (link instanceof MandatoryModIdDefinition) {
+							MandatoryModIdDefinition mandatoryMod = (MandatoryModIdDefinition) link;
+							roots.put(mandatoryMod.candidate, mandatoryMod);
+							iterator.remove();
 						}
 					}
-				}
 
-				//noinspection UnnecessaryLocalVariable
-				IProblem problem = solver;
-				IVecInt assumptions = new VecInt(modCandidateMap.size());
+					// Plugin functionality to handle errors would likely go *RIGHT HERE*
+					// I.E. downloading missing / outdated *optional* mods
+
+					/*
+					 * Plugins would most likely be passed the map of roots, and the list of causes.
+					 * Alternatively, we could do some pre-processing to identify specific cases, such as
+					 * a mod or library (or api class implementation) being missing, and then send those
+					 * out separately.
+					 *
+					 * This does mean that we'd need to remove and re-add edited clauses though when
+					 * trying to fix problems. (For example downloading a mod should process all of the
+					 * dependencies, provides, etc of that mod related to all others). 
+					 */
+
+					ModResolutionException ex = describeError(roots, causes);
+					if (ex == null) {
+						ex = fallbackErrorDescription(roots, causes);
+					}
+
+					errors.add(ex);
+
+					if (causes.isEmpty()) {
+						break;
+					} else {
+
+						boolean removedAny = false;
 
-				for (String mod : modCandidateMap.keySet()) {
-					int pos = assumptions.size();
-					assumptions = assumptions.push(0);
-					Collection<ModCandidate> candidates = modCandidateMap.get(mod);
-					boolean satisfied = false;
+						// Remove dependences and conflicts first
+						for (ModLink link : causes) {
 
-					for (ModCandidate candidate : candidates) {
-						assumptions.set(pos, candidateIntMap.get(candidate));
-						if (problem.isSatisfiable(assumptions)) {
-							satisfied = true;
+							if (link instanceof ModDep) {
+								ModDep dep = (ModDep) link;
+
+								if (!dep.validOptions.isEmpty()) {
+									continue;
+								}
+							}
+
+							if (link instanceof ModDep || link instanceof ModBreakage) {
+								if (helper.quilt_removeConstraint(link)) {
+									removedAny = true;
+									break;
+								}
+							}
+						}
+
+						// If that failed... try removing anything else
+						if (!removedAny) {
+							for (ModLink link : causes) {
+								if (helper.quilt_removeConstraint(link)) {
+									removedAny = true;
+									break;
+								}
+							}
+						}
+
+						// If that failed... stop finding more errors
+						if (!removedAny) {
 							break;
 						}
 					}
+				}
 
-					if (!satisfied) {
-						if (mandatoryMods.contains(mod)) {
-							throw new ModResolutionException("Could not resolve mod collection including mandatory mod '" + mod + "'");
-						} else {
-							assumptions = assumptions.pop();
-						}
+				if (!errors.isEmpty()) {
+					if (errors.size() == 1) {
+						throw errors.get(0);
```

Comment:
May be worth wrapping the error at idx = 0 in a mod resolution exception as well so we don't lose the description like below?

Reply:
Erm, maybe? If there's only a single error then I can't see why we'd need to say explicitly that there's only 1 error. (This is also to ensure it gets displayed in the GUI correctly - otherwise we'd end up with different errors at different indentation levels. However if there's only 1 error then there's only 1 indentation level to worry about - so we might as well *not* include a header error).